proxy/src/rate_limiter/limiter.rs

```diff
@@ -15,6 +18,69 @@ use super::{
     RateLimiterConfig,
 };
 
+// Simple per-endpoint rate limiter.
+//
+// Check that number of connections to the endpoint is below `max_rps` rps.
+// Purposefully ignore user name and database name as clients can reconnect
+// with different names, so we'll end up sending some http requests to
+// the control plane.
+//
+// TODO: add a better bucketing here, e.g. not more than 300 requests per second,
+//       and not more than 1000 requests per 10 seconds, etc. Short bursts of reconnects
+//       are noramal during redeployments, so we should not block them.
+pub struct EndpointRateLimiter {
+    map: DashMap<SmolStr, Arc<Mutex<(chrono::NaiveTime, u32)>>>,
+    max_rps: u32,
+    access_count: AtomicUsize,
```

Comment:
```suggestion
    access_count: AtomicU64,
```

nit: Realistically doesn't matter, but iiuc u64 is more correct

Reply:
hmm, why? on x64 is would be the same size, but on x32 it would be 4 bytes, but atomic (atomic u64 will be wrapped with some mutex most likely).